garak/generators/websocket.py

```diff
@@ -0,0 +1,379 @@
+"""WebSocket generator for real-time LLM communication
+
+This module provides WebSocket-based connectivity for testing LLM services
+that use real-time bidirectional communication protocols.
+"""
+
+import asyncio
+import json
+import time
+import base64
+import os
+import logging
+from typing import List, Union, Dict, Any, Optional
+from urllib.parse import urlparse
+import websockets
+from websockets.exceptions import ConnectionClosed, WebSocketException
+
+from garak import _config
+from garak.attempt import Message, Conversation
+from garak.generators.base import Generator
+
+logger = logging.getLogger(__name__)
+
+
+class WebSocketGenerator(Generator):
+    """Generator for WebSocket-based LLM services
+    
+    This generator connects to LLM services that communicate via WebSocket protocol,
+    handling authentication, template-based messaging, and JSON response extraction.
+    
+    Configuration parameters:
+    - uri: WebSocket URL (ws:// or wss://)
+    - name: Display name for the service
+    - auth_type: Authentication method (none, basic, bearer, custom)
+    - username: Basic authentication username
+    - api_key: API key for bearer token auth or password for basic auth
+    - req_template: String template with $INPUT and $KEY placeholders
+    - req_template_json_object: JSON object template for structured messages
+    - headers: Additional WebSocket headers
+    - response_json: Whether responses are JSON formatted
+    - response_json_field: Field to extract from JSON responses (supports JSONPath)
+    - response_after_typing: Wait for typing indicator completion
+    - typing_indicator: String that indicates typing status
+    - request_timeout: Seconds to wait for response
+    - connection_timeout: Seconds to wait for connection
+    - max_response_length: Maximum response length
+    - verify_ssl: SSL certificate verification
+    """
+
+    DEFAULT_PARAMS = {
+        "uri": "wss://echo.websocket.org",
+        "name": "WebSocket LLM",
```

Comment:
this needs to come out and become a class-level immutable attribute. running the tests will indicate the current failure